src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1147,9 +1149,47 @@ async ValueTask<NpgsqlDataReader> ExecuteReaderAsync(CommandBehavior behavior, b
                 if (conn.TryGetBoundConnector(out var connector))
                 {
                     connector.StartUserAction(this);
+
+                    CancellationTokenRegistration? registration = null;
+
                     try
                     {
-                        using var _ = cancellationToken.Register(cmd => ((NpgsqlCommand)cmd!).Cancel(), this);
+                        var finalCt = CancellationToken.None;
+
+                        if (cancellationToken.CanBeCanceled)
+                        {
+                            _boundConnector = connector;
+
+                            if (connector.CommandCts.IsCancellationRequested)
+                            {
+                                connector.CommandCts.Dispose();
+                                connector.CommandCts = new CancellationTokenSource();
+                            }
+
+                            registration = cancellationToken.Register(o =>
+                            {
+                                var cmd = (NpgsqlCommand)o!;
+                                var cn = cmd._boundConnector!;
+                                lock (cn)
```

Comment:
What exactly is this lock supposed to protect against? Since cancellation works by opening a new physical connection, it's actually thread-safe with anything happening on the (main) connection.

Reply:
I've misunderstood your comment from the last review. I've taken another look, and it should be fine now.